### PR TITLE
294: Use and re-export fhirtypes package in core

### PR DESCRIPTION
**Pull request by ianatacn**
## Why
https://github.com/bonfhir/bonfhir/issues/294

Use @bonfhir/fhirtypes types in the core package

## What's changing
- Import r4b/r5 fhir types from @bonfhir/fhirtypes as needed for core `src`
- Dynamically import types from @bonfhir/fhirtypes in templates
- Re-export generated fhir types for backwards compatibility

### CODE REVIEW

File: packages/core/rollup.config.mjs [hunk 1]
```diff
@@ -74,6 +74,7 @@ export default ["r4b", "r5"].flatMap((release) =>
           format: format === "esm" ? "m" : "c",
         },
       ],
+      external: ['@bonfhir/fhirtypes/r4b', '@bonfhir/fhirtypes/r5'],
```
Comment (ianatacn) [inline packages/core/rollup.config.mjs]: @opiation marking these dependencies as external resolves the warning but at the same time, we _want_ to include the types from fhirtypes in the bundle. 

They are included as part of the js bundle however, so I wanted to get your opinion here. Thank you!

File: packages/core/src/r4b/index.ts [hunk 2]
```diff
@@ -1,3 +1,4 @@
+export * from "@bonfhir/fhirtypes/r4b";
 export * from "./builders";
```
Comment (ianatacn) [inline packages/core/src/r4b/index.ts:2]: Export all from fhirtypes package for backwards compatibility

File: packages/core/rollup.config.mjs (hunk 1 above)
Comment (opiation) [reply packages/core/rollup.config.mjs]: I'm confused. My understanding of Rollup's [`external` config]() is that packages listed there are _not_ bundled with the code being _rolled up_. Sounds to me like that means that `@bonfhir/core` won't include the types from `@bonfhir/fhirtypes` that it re-exports. You mention though that it _is_ included. How did you validate that? Essentially, we _should_ be able to import something like `Patient` from the generated `@bonfhir/core` bundle like:

```ts
// in /packages/core/import_test.ts

import type { Patient } from "./dist/r4b/esm/index";
//            ^^^ Should not be `any`
```

**Review by opiation — APPROVED** [review]
I'm approving tentatively on the grounds that the `@bonfhir/core` must re-export the `@bonfhir/fhirtypes` so they're importable from `@bonfhir/core` even in projects that don't explicitly include `@bonfhir/fhirtypes` (as per the relevant comment).

File: packages/core/rollup.config.mjs (hunk 1 above)
Comment (ianatacn) [reply packages/core/rollup.config.mjs]: @opiation the key thing to note is that the `input` of the ts rollup config is `dist/${release}/${format}/dts/index.d.ts` .. the output of the js rollup config that _does_ include bonfhir/fhirtypes. I will look into adding a test suite in core to explicitly validate the re-exported types

Comment (ianatacn) [reply packages/core/rollup.config.mjs]: Tested and confirmed that functions and re-exported types are loaded correctly. So we are good to merge 👍

**Comment by opiation** [comment]
For the sake of documentation and as per our conversation, we confirmed that with the changes submitted here, runtime values exported by `@bonfhir/fhirtypes`  are re-exported by `@bonfhir/core`. Namely, the `isResource` function and the `DomainResourceTypes` array are inlined in the `@bonfhir/core` r4 and r5 CJS bundles. The types are re-exported which is sufficient since any consumer of the `@bonfhir/core` library will have the `@bonfhir/fhirtypes` package installed as a transitive dependency.

**Pull request merged**